crates/pathfinder/src/update.rs

```diff
@@ -0,0 +1,170 @@
+                    .unwrap_or(&release.version);
+                match semver::Version::parse(release_version) {
+                    Ok(version) => {
+                        latest_gh_version = Some(version);
+                    }
+                    Err(e) => {
+                        tracing::warn!(error=%e, version=%release.version, "Semver parsing of latest github release failed")
+                    }
+                };
+            }
+            UpdateResult::NotModified => {
+                tracing::trace!(latest=?latest_gh_version, "No new release found on Github");
+            }
+            UpdateResult::RewestError(e) if e.is_decode() || e.is_body() || e.is_builder() => {
+                // More severe errors, probably indicating something is wrong with our setup.
+                // Set to warn and not error because this update checking is a non-critical feature.
+                tracing::warn!(error=%e, "Error checking Github for new releases")
+            }
+            UpdateResult::RewestError(e) => {
+                // Less severe errors, includes transient connection errors and timeouts; does not warrant
+                // a high log level.
+                tracing::trace!(error=%e, "Error checking Github for new releases")
+            }
+            UpdateResult::Other(e) => {
+                // Other, unexpected errors.
+                tracing::warn!(error=%e, "Error checking Github for new releases");
+            }
+        }
+
+        // Display latest release info if it is newer than this application.
+        if let Some(github) = latest_gh_version.as_ref() {
+            if github > &local_version {
+                tracing::warn!(release=%github, "New pathfinder release available! Please consider updating your node!")
+            }
+        }
+
+        // Only poll every 10 minutes.
+        tokio::time::sleep(std::time::Duration::from_secs(60 * 10)).await;
+    }
+}
+
+#[derive(Debug)]
+struct Release {
+    version: String,
+    /// Optional because its possible to have an invalid string ETag header.
+    etag: Option<String>,
+}
+
+#[derive(Debug)]
+enum UpdateResult {
+    /// A new latest release.
+    Update(Release),
+    /// No new releases since last query (status code 304).
+    NotModified,
+    RewestError(reqwest::Error),
+    Other(anyhow::Error),
+}
+
+/// Fetches the latest pathfinder [Release] from Github using their REST API.
+///
+/// The [IF_NONE_MATCH](reqwest::header::IF_NONE_MATCH) header options is set to
+/// the `etag` parameter to prevent Github from sending redundant information. The
+/// resulting 304 status code is mapped to [UpdateResult::NotModified].
+async fn fetch_latest_github_release(
+    client: &reqwest::Client,
+    etag: &Option<String>,
+) -> UpdateResult {
+    use reqwest::StatusCode;
+    use reqwest::Url;
+
+    const USER_AGENT: &str = concat!(
+        env!("CARGO_PKG_NAME"),
+        "/",
+        env!("VERGEN_GIT_SEMVER_LIGHTWEIGHT")
+    );
```

Comment:
Copypasting this around kind of makes me think it could be in `lib.rs` or `mod consts`, since it's non-trivial and I have a feeling that vergen part might need to be changed in future :) But not a huge deal.

Will this be pathfinder_lib/vergen btw? If so, is that intentional?

Reply:
> Copypasting this around kind of makes me think it could be in lib.rs or mod consts, since it's non-trivial and I have a feeling that vergen part might need to be changed in future :) But not a huge deal.
Not a bad idea.

> Will this be pathfinder_lib/vergen btw? If so, is that intentional?
No, e.g.: `USER_AGENT = "pathfinder/v0.1.8-alpha-5-g115b28f"`. It takes the package name from the `cargo.toml` manifest which currently is:
```
[package]
name = "pathfinder"
```

We can maybe argue that we should be using the executable name for this though.

Reply:
Oki this is good for me if it's "pathfinder/..."